### PR TITLE
Add 'save_file' kwargs for GUROBI.

### PR DESCRIPTION
**Pull request by QiuWJX**
- use Gurobi.write(save_file) to save the model file.
- cplex and mosek have the similar functionality.

## Description
Very simple changes.
Make it possible to write model using gurobi.

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [ ] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [ ] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: cvxpy/reductions/solvers/conic_solvers/gurobi_conif.py [hunk 1]
```diff
@@ -212,6 +212,11 @@ def solve_via_data(self, data, warm_start: bool, verbose: bool, solver_opts, sol
             variables += new_vars
             soc_start += constr_len
 
+        # Save file (*.mst, *.sol, ect.)
+        if 'save_file' in solver_opts:
+            model.write(solver_opts['save_file'])
+            del solver_opts['save_file']
```
Comment (SteveDiamond) [inline cvxpy/reductions/solvers/conic_solvers/gurobi_conif.py:218]: No need to delete.

**Comment by SteveDiamond** [comment]
Thanks for adding this! Could you write a test as well? It's not so much to validate the code (which is very simple), as to commit us to maintaining this functionality.

**Comment by QiuWJX** [comment]
A test is added

**Comment by SteveDiamond** [comment]
Thanks for adding the test! The CI is failing because of code style issues. Please follow these instructions: https://www.cvxpy.org/contributing/index.html#code-style

@rileyjmurray how should this test be integrated into the solver tests?

**Comment by QiuWJX** [comment]
- the code style has been changed
- the gurobi_model.lp has been moved to /resources/ folder
- BUT I got the error "ImportError: cannot import name 'InverseData' from partially initialized module 'cvxpy.reductions' (most likely due to a circular import) (g:\codes\python\cvxpy\cvxpy\reductions\__init__.py)". I thought it is for the change of code style . I have no idea about whether it is an error only happens in my PC. Could you help me?

**Comment by SteveDiamond** [comment]
No worries, we'll take it from here.

File: cvxpy/tests/test_gurobi_write.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+"""
+Copyright 2022 Jiaxin Wang
```
Comment (rileyjmurray) [inline cvxpy/tests/test_gurobi_write.py:2]: @QiuWJX please change this line to "Copyright 2022, the CVXPY authors"

File: cvxpy/tests/test_gurobi_write.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import os
+
+import numpy as np
+
+import cvxpy as cp
+from cvxpy.tests.base_test import BaseTest
+
+
+class TestGurobiWrite(BaseTest):
+    """ Unit tests for the expression/expression module. """
```
Comment (rileyjmurray) [inline cvxpy/tests/test_gurobi_write.py:25]: @QiuWJX this docstring isn't correct

**Pull request merged**